core/src/main/scala/chisel3/Module.scala

```diff
@@ -406,7 +406,8 @@ package experimental {
     }
 
     /** Legalized name of this module. */
-    final lazy val name =
+    final lazy val name: String = {
+      _parent.map(_.name) // Name parent before me!
```

Comment:
I'll add a test to double-check this.

Reply:
This is definitely a problem. 🥲 Note: there is not null pointer exception, but the `name` will be `null`. If downstream things don't know how to handle this they may throw exceptions.

This can't be made into a `def` because it does need to be a single value, too.

The approach here seems like it should work. Is there any problem if `val incremental` is declared after the submodule is instantiated, e.g.:

```
class Bar extends Module {}
class Foo extends Module {
  val bar = Module(new Bar)
  val incremental = n + "Wrapper"
  override def desiredName = incremental
}
```

Reply:
> Is there any problem if val incremental is declared after the submodule is instantiated

Yes, it will have the same error, but the fix is possible (move the val earlier). I'll update the error message to suggest this.